src/LogExpert.Core/Classes/Log/LogfileReader.cs

```diff
@@ -755,43 +779,74 @@ private ILogFileInfo AddFile (string fileName)
         return info;
     }
 
-    private Task<ILogLine> GetLogLineInternal (int lineNum)
+    private Task<ILogLine> GetLogLineInternal (int lineNum, CancellationToken ct)
+    {
+        // Run all the heavy work off the UI thread,
+        // and honor the cancellation token
+        return Task.Run(() => GetLogLineInternalCore(lineNum, ct), ct);
+    }
+
+    private ILogLine GetLogLineInternalCore (int lineNum, CancellationToken ct)
     {
+        ct.ThrowIfCancellationRequested();
+
         if (_isDeleted)
         {
-            _logger.Debug(CultureInfo.InvariantCulture, "Returning null for line {0} because file is deleted.", lineNum);
+            _logger.Debug(
+                CultureInfo.InvariantCulture,
+                "Returning null for line {0} because file is deleted.",
+                lineNum);
 
-            // fast fail if dead file was detected. Prevents repeated lags in GUI thread caused by callbacks from control (e.g. repaint)
             return null;
         }
 
-        AcquireBufferListReaderLock();
-        LogBuffer logBuffer = GetBufferForLine(lineNum);
-        if (logBuffer == null)
+        AcquireBufferListReaderLock();    // blocking call off UI
+        try
         {
-            ReleaseBufferListReaderLock();
-            _logger.Error("Cannot find buffer for line {0}, file: {1}{2}", lineNum, _fileName, IsMultiFile ? " (MultiFile)" : "");
-            return null;
-        }
+            ct.ThrowIfCancellationRequested();
 
-        // disposeLock prevents that the garbage collector is disposing just in the moment we use the buffer
-        _disposeLock.AcquireReaderLock(Timeout.Infinite);
-        if (logBuffer.IsDisposed)
-        {
-            LockCookie cookie = _disposeLock.UpgradeToWriterLock(Timeout.Infinite);
-            lock (logBuffer.FileInfo)
+            var logBuffer = GetBufferForLine(lineNum);
+            if (logBuffer == null)
             {
-                ReReadBuffer(logBuffer);
+                _logger.Error("Cannot find buffer for line {0}, file: {1}{2}", lineNum, _fileName, IsMultiFile ? " (MultiFile)" : "");
+                return null;
             }
 
-            _disposeLock.DowngradeFromWriterLock(ref cookie);
-        }
+            _disposeLock.AcquireReaderLock(Timeout.Infinite);
+            try
+            {
+                ct.ThrowIfCancellationRequested();
```

Comment:
I know it's just a draft, but the try, finally is imo worse then the old design

Also I'm confused why so many ct.ThrowIfCancellationRequested();